server/src/main/java/org/apache/iotdb/db/mpp/buffer/SinkHandle.java

```diff
@@ -0,0 +1,346 @@
+
+  private void sendEndOfDataBlockEvent() throws TException {
+    logger.debug(
+        "Send end of data block event to operator {} of {}.",
+        remoteOperatorId,
+        remoteFragmentInstanceId);
+    int attempt = 0;
+    EndOfDataBlockEvent endOfDataBlockEvent =
+        new EndOfDataBlockEvent(
+            remoteFragmentInstanceId,
+            remoteOperatorId,
+            localFragmentInstanceId,
+            bufferedTsBlocks.size() - 1);
+    while (attempt < MAX_ATTEMPT_TIMES) {
+      attempt += 1;
+      try {
+        client.onEndOfDataBlockEvent(endOfDataBlockEvent);
+        break;
+      } catch (TException e) {
+        logger.error(
+            "Failed to send end of data block event to operator {} of {} due to {}, attempt times: {}",
+            remoteOperatorId,
+            remoteFragmentInstanceId,
+            e.getMessage(),
+            attempt);
+        if (attempt == MAX_ATTEMPT_TIMES) {
+          throw e;
+        }
+      }
+    }
+  }
+
+  @Override
+  public void close() throws IOException {
+    logger.info("Sink handle {} is being closed.", this);
+    if (throwable != null) {
+      throw new IOException(throwable);
+    }
+    if (closed) {
+      return;
+    }
+    synchronized (this) {
+      closed = true;
+      noMoreTsBlocks = true;
+    }
+    sinkHandleListener.onClosed(this);
+    try {
+      sendEndOfDataBlockEvent();
+    } catch (TException e) {
+      throw new IOException(e);
+    }
+    logger.info("Sink handle {} is closed.", this);
+  }
+
+  @Override
+  public void abort() {
+    logger.info("Sink handle {} is being aborted.", this);
+    synchronized (this) {
+      bufferedTsBlocks.clear();
+      numOfBufferedTsBlocks = 0;
+      closed = true;
+      localMemoryManager
+          .getQueryPool()
+          .free(localFragmentInstanceId.getQueryId(), bufferRetainedSizeInBytes);
+      bufferRetainedSizeInBytes = 0;
+    }
+    sinkHandleListener.onAborted(this);
+    logger.info("Sink handle {} is aborted", this);
+  }
+
+  @Override
+  public synchronized void setNoMoreTsBlocks() {
+    noMoreTsBlocks = true;
+  }
+
+  @Override
+  public boolean isClosed() {
+    return closed;
+  }
+
+  @Override
+  public boolean isFinished() {
+    return throwable == null && noMoreTsBlocks && numOfBufferedTsBlocks == 0;
+  }
+
+  @Override
+  public long getBufferRetainedSizeInBytes() {
+    return bufferRetainedSizeInBytes;
+  }
+
+  @Override
+  public int getNumOfBufferedTsBlocks() {
+    return numOfBufferedTsBlocks;
+  }
+
+  ByteBuffer getSerializedTsBlock(int partition, int sequenceId) {
+    throw new UnsupportedOperationException();
+  }
+
+  ByteBuffer getSerializedTsBlock(int sequenceId) {
+    TsBlock tsBlock;
+    synchronized (this) {
+      tsBlock = bufferedTsBlocks.get(sequenceId);
+      if (tsBlock == null) {
+        throw new IllegalStateException("The data block doesn't exist. Sequence ID: " + sequenceId);
+      }
+      bufferedTsBlocks.set(sequenceId, null);
```

Comment:
Setting null here is not safe, what if this response failed, so you can try the above way which can avoid this.

Reply:
Fixed.